src/pyodide.js

```diff
@@ -171,102 +161,127 @@ var languagePluginLoader = new Promise((resolve, reject) => {
       return baseURL + path;
     };
 
-    let promise = new Promise((resolve, reject) => {
-      if (Object.keys(toLoad).length === 0) {
-        resolve('No new packages to load');
-        return;
+    if (Object.keys(toLoad).length === 0) {
+      return Promise.resolve('No new packages to load');
+    } else {
+      console.log(`Loading ${Object.keys(toLoad).join(', ')}`)
+    }
+
+    // When we load a package, we first load a JS file, and then the JS file
+    // loads the data file. The status of the first step is tracked by the
+    // loadScript function above. The status of the second step is tracked by
+    // emscripten's runDependency system.
+    //
+    // When the JS file is loaded, it *synchronously* adds a runDependency to
+    // emscripten. When it finishes loading, it removes the runDependency.
+    // Whenever the number of runDependencies changes, monitorRunDependency is
+    // called with the number of pending dependencies.
+    //
+    // Under this system, packages have completed loading if all javascript
+    // files are loaded, and then the number of runDependencies becomes zero.
+    // The number of runDependencies may happen to equal zero in between
+    // package files loading if we are loading multiple packages. We must avoid
+    // accidentally returning early if this happens.
+
+    // runDepPromise is a promise such that at any point in time, it is
+    // resolved if and only if the number of pending run dependencies is 0.
+    // We assume that at this point in time, there are no pending run
+    // dependencies.
+    //
+    // The promise it refers to may be replaced when a runDependency is added,
+    // so we must be careful to only await for this after we know all
+    // runDependencies have been added.
+    let runDepPromise = Promise.resolve();
+    // runDepResolve is the resolve function of runDepPromise if it is
+    // unresolved, undefined otherwise.
+    let runDepResolve = undefined;
+
+    // We must start monitoring before we load any script, as this function is
+    // triggered by the script loading.
+    self.pyodide._module.monitorRunDependencies = (n) => {
+      if (n === 0) {
+        if (runDepResolve !== undefined) {
+          runDepResolve();
+          runDepResolve = undefined;
+        }
+      } else {
+        if (runDepResolve === undefined) {
+          runDepPromise = new Promise(r => runDepResolve = r);
+        }
       }
+    };
 
-      let packageList = Array.from(Object.keys(toLoad));
-      _messageCallback(`Loading ${packageList.join(', ')}`)
+    // Add a promise that rejects when any exceptions is thrown in the process
+    // of loading a script. The promise never resolves, and we combine it
+    // with other promises via Promise.race.
+    let windowErrorHandler;
+    let windowErrorPromise = new Promise((_res, rej) => {
+      windowErrorHandler = e => rej(e.message);
```

Comment:
```suggestion
      windowErrorHandler = e => {
         console.error("Unhandled error. We don't know what it is or whether it is related to 'loadPackage' but out of an abundance of caution we will assume that loading failed.");
         console.error(e);
         rej(e.message);
      };
```
We're logging info level stuff all over the place, but the one time we grab any and all unhandled errors regardless of whether they have anything to do with us and we don't even log it!